### PR TITLE
Clams routes and schema creation

### LINKED ISSUE #5: create clams SQL table and CRUD backend routes
Description under "Backend Route Groups"
https://docs.google.com/document/d/15oq7y1-zg20i2gDJD0T31X9wJwE--TKntyKFzHCKW5E/edit?usp=sharing 

Here's [how to create a table](https://www.tutorialspoint.com/mysql/mysql-create-tables.html) in mysql.
Put this sql file under a folder /server/schema (ex: /server/schema/user.sql)

### PR DESCRIPTION
**Pull request by sravtall**
#5 

Let us know if any changes are required. Thanks.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -3,6 +3,8 @@ const cors = require('cors');
 
 require('dotenv').config();
 
+const clams = require('./routes/clams');
```
Comment (alanchangxyz) [inline app.js]: let's call this `clamsRouter` so we can distinguish what we're importing from `clams.js` better

File: server/schema/clams.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE clams IF NOT EXISTS{
```
Comment (alanchangxyz) [inline server/schema/clams.sql]: the `clams` table hasn't been made in the database yet; it should be in there for this task to be finished

File: server/schema/clams.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    clamId SERIAL PRIMARY KEY,
+    rakerId INTEGER NOT NULL,
+    lat DECIMAL NOT NULL,
+    long DECIMAL NOT NULL,
+    length INTEGER NOT NULL,
```
Comment (alanchangxyz) [inline server/schema/clams.sql]: `length`, `width`, `weight` might be better off as decimals as opposed to integers

File: routes/clams.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  try {
+    const clam = await pool.query('SELECT * FROM clams WHERE clamId = $(clamId);');
+    res.status(200).json(keysToCamel(clam.rows));
+  } catch (err) {
+    res.status(400).send(err.message);
+  }
+});
+
+// Get data for clams based on rakerId
+router.get('/rakers/:rakerId/clams', async (req, res) => {
```
Comment (alanchangxyz) [inline routes/clams.js]: this route scheme is a little long, since we're getting a raker's clams from `/clams` already you may be able to shorten this to something like `/clams/raker/:rakerId`

File: routes/clams.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  try {
+    const clam = await pool.query('SELECT * FROM clams WHERE rakerId = $(rakerId)');
```
Comment (alanchangxyz) [inline routes/clams.js]: an object should be passed in to the query with your `rakerId` you're planning to insert into the query (see other backends' queries for examples)

tldr just test this route after you fix it and add the clams table

File: routes/clams.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    res.status(400).send(err.message);
+  }
+});
+
+// Create clam
+router.post('/', async (req, res) => {
+  try {
+    const { clamId, rakerId, lat, long, length, width, weight, comments, image } = req.body;
+    const clam = await pool.query(
+      `INSERT INTO clams (
```
Comment (alanchangxyz) [inline routes/clams.js]: there's no sign of the clam attributes being passed into the query here; see my comment a little above on how to best fix it

tldr: more testing

File: routes/clams.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    res.status(400).send(err.message);
+  }
+});
+
+// Delete clam
+router.delete('/:clamId', async (req, res) => {
+  try {
+    const { clamId } = req.params;
+    const deletedClam = await pool.query(
+      `DELETE from clams WHERE clamId = $(clamId) RETURNING *;`,
```
Comment (alanchangxyz) [inline routes/clams.js]: revision/testing (see above)

File: routes/clams.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+// Update clam
+router.put('/', async (req, res) => {
+  try {
+    const { clamId } = req.params;
+    const { rakerId, lat, long, length, width, weight, comments, image } = req.body;
+    const updatedClam = await pool.query(
+      `UPDATE clam
+         SET
+         ${clamId ? 'clamId, ' : ''}
```
Comment (alanchangxyz) [inline routes/clams.js]: need to pass in attributes via an object (see above)

**Review by alanchangxyz — DISMISSED** [review]
it's a good start but we're gonna need some more work here, mysql workbench is your best friend for viewing and querying the db as you implement them here (if you have trouble using it, just ask)

**Comment by alanchangxyz** [comment]
OK ignore my comments related to passing attributes in via an object; a lot more things need to be fixed. @Jacqueline1201 @sirAvent can explain in standup this Tuesday

File: server/schema/clams.sql [hunk 9]
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE clam IF NOT EXISTS{
+    clamId INTEGER PRIMARY KEY,
```
Comment (alanchangxyz) [inline server/schema/clams.sql]: in a sql table we should use `underscore_case` for fields instead of `camelCase`

`keysToCamel` is the util function on the express backend that will convert these fields to `camelCase` for us when we return stuff from queries

please update the db to reflect the new schema too afterwards (hint: `UPDATE TABLE` statements) + any corresponding routes that added/changed/referenced `rakerId`

File: routes/clams.router.js [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    const clam = await pool.query(query, params);
+
+    res.status(200).json(keysToCamel(clam));
+  } catch (err) {
+    res.status(400).send(err.message);
+  }
+});
+
+// Delete clam
+router.delete('/:clamId', async (req, res) => {
```
Comment (alanchangxyz) [inline routes/clams.router.js]: [nit] `delete` routes should ideally be placed after `put` routes in CRUD order; let's switch the placements in this file to reflect that

File: routes/clams.router.js [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+  }
+});
+
+// Update clam
+router.put('/:clamId', async (req, res) => {
+  try {
+    const { clamId } = req.params;
+    isNumeric(clamId);
+
+    const { rakerId, lat, lon, length, width, weight, comments, image } = req.body;
```
Comment (alanchangxyz) [inline routes/clams.router.js:109]: `lat`, `lon`, `length`, `width`, `weight` can be checked with `isNumeric` too here since they are all numbers

edit: on the `POST` route too

File: app.js [hunk 12]
```diff
@@ -12,7 +14,11 @@ app.use(
     origin: `${process.env.REACT_APP_HOST}:${process.env.REACT_APP_PORT}`,
   }),
 );
+app.use(express.json());
+
+app.use('/clams', clamsRouter);
 
+// eslint-disable-line
```
Comment (alanchangxyz) [inline app.js]: although it is probably annoying when eslint gets mad at the `console.log` line here let's keep the warning instead of adding an unnecessary line disable

File: routes/clams.router.js [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+         SET
+         ${rakerId ? 'rakerId = :rakerId, ' : ''}
+         ${lat ? 'lat = :lat, ' : ''}
+         ${lon ? 'lon = :lon, ' : ''}
+         ${length ? 'length = :length, ' : ''}
+         ${width ? 'width = :width, ' : ''}
+         ${weight ? 'weight = :weight, ' : ''}
+         ${comments ? 'comments = :comments, ' : ''}
+         ${image ? 'image = :image, ' : ''}
+        clamId = :clamId
```
Comment (alanchangxyz) [inline routes/clams.router.js]: [nit] random odd tabspace here

**Review by alanchangxyz — COMMENTED** [review]
i'll functional test this later but adding syntax comments now:

File: routes/clams.router.js [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    const clam = await pool.query(query, params);
+
+    res.status(200).json(keysToCamel(clam));
+  } catch (err) {
+    res.status(400).send(err.message);
+  }
+});
+
+// Get data for clams based on rakerId
+router.get('/rakers/:rakerId', async (req, res) => {
```
Comment (alanchangxyz) [inline routes/clams.router.js]: [nit] this might be better off as `/raker/:rakerId`

File: routes/clams.router.js [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+         ${rakerId ? 'rakerId = :rakerId, ' : ''}
+         ${lat ? 'lat = :lat, ' : ''}
+         ${lon ? 'lon = :lon, ' : ''}
+         ${length ? 'length = :length, ' : ''}
+         ${width ? 'width = :width, ' : ''}
+         ${weight ? 'weight = :weight, ' : ''}
+         ${comments ? 'comments = :comments, ' : ''}
+         ${image ? 'image = :image, ' : ''}
+        clamId = :clamId
+         WHERE clamId = :clamId;`,
```
Comment (alanchangxyz) [inline routes/clams.router.js]: we'll have to add an additional sql statement to get the newly edited clam's info after we edit it here, otherwise it'll just give us the results of the prev query which aren't that useful for us (if you look at the route results right now it's just a bunch of sql junk)

UPDATE: see my comment above on the results from multiple sql queries in `clam`

File: routes/clams.router.js [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        width,
+        weight,
+        comments,
+        image,
+      },
+    );
+
+    const clam = await pool.query(query, params);
+
+    res.status(200).json(keysToCamel(clam));
```
Comment (alanchangxyz) [inline routes/clams.router.js]: okay after running the results from this line i think we'll need to get the second result from `clam` since it contains multiple queries' results as a list; the second object in the list should be the clam info we need from the `SELECT` statement here

**Review by claudeisakeeb — APPROVED** [review]

**Review by Jacqueline1201 — APPROVED** [review]
